### ISSUE TITLE
Make all immutable interfaces `sealed` to prevent implementations that allow mutations

### ISSUE THREAD

**Issue by yogurtearl** [Kotlin/kotlinx.collections.immutable#147]
Right now you can roll your own mutable version of `ImmutableList` with something like this:

```kotlin
class SneakyList<E>(
    val mutableList: MutableList<E> = mutableListOf()
): ImmutableList<E> {
    override val size: Int
        get() = mutableList.size

    override fun get(index: Int): E = mutableList[index]
.
.
.
}
```

Even well-meaning implementations of the `ImmutableList` interface might accidentally enable mutability.
To make `ImmutableList` (and `ImmutableCollection`, etc) even more immutable, make all of the the interfaces in this library `sealed` interfaces.

**Comment by wzsaz** [comment]
I don't think this is necessary. Wrongfully implementing an interface will always cause problems.
Just because someone could implement the List interface incorrectly you would not make it sealed.
Sealed interfaces should be conceptually reserved to represent closed implementation hierarchies.

**Comment by yogurtearl** [comment]
If the implementation of the interface can be mutable, then it is not "immutable" it is just a readonly interface similar to `List`

**Comment by Laxystem** [comment]
> If the implementation of the interface can be mutable, then it is not "immutable" it is just a readonly interface similar to `List`

Correct. But immutability is a contract. Just like Java's mutability is a contract broken by Guava's immutable collections. Making `ImmutableList` sealed will prevent people from implementing immutable collections on their own.

Consider this:

https://github.com/Kotlin/kotlinx.collections.immutable/blob/1d18389e32e8afdbf5caa89808fb5af308fb5c8d/core/commonMain/src/adapters/ReadOnlyCollectionAdapters.kt#L11C1-L20C2

**Comment by qurbonzoda** [comment]
The `Immutable` and `Persistent` interfaces were introduced to allow users implement alternative versions that better suite their use cases. For example, in projects where collections are frequently accessed but very rarely updated, a regular copy-on-write implementation of these interfaces might be more appropriate. However, we recognize that only a small number of users might find this flexibility useful. Therefore, it is very helpful to hear from you whether you utilize this flexibility or it instead poses challenges for you.

In the future, we might decide to limit the implementation of these interfaces or expose the implementation classes. So that users can be confident that the implementation is indeed immutable.

**Comment by yogurtearl** [comment]
I find that `kotlin.collections.List` provides the needed flexibility for alternative list implementations. 
But when using `ImmutableList` I am looking for the strongest possible multiplatform immutability guarantees from a security and thread safety point of view. (I know reflection, serialization, agents, etc can be used to subvert some guarantees)

As an example, if I bring in 2 third party SDKs (sdk1 and sdk2) I want to be sure that `immutableList` won't change after I pass it into sdk2:

```kotlin
val immutableList: ImmutableList = sdk1.getData()

val processedData: List = sdk2.process(immutableList)

// immutableList is unchanged
```

**Comment by Laxystem** [comment]
To further my point, if we make `ImmutableList` `sealed`, shouldn't we do the same for mutable ones, too? Afterall, the same arguments apply - the contract can be broken, it could have some performance benefits...

What you're suggesting applies to any interface: if we couple the implementation with the API (by making the interface sealed), we can get better performance and reliability.

**Comment by Sporking** [comment]
I don't think ImmutableList should be a sealed interface. I think it should be a final class, not an interface, which is what Google Guava does. Guava does it that way for pretty much exactly the reasons described in this bug report. (Allowing additional implementations of ImmutableList and other immutable types can create serious security holes.)

This is not the first time this issue has been brought up. See for instance my comments [here](https://github.com/Kotlin/kotlinx.collections.immutable/issues/10) which was filed in 2017.

**Comment by Laxystem** [comment]
@Sporking about security: in all Kotlin targets, the moment malicious code runs, you're fucked. Unless JetBrains is planning on improving security across the board in ways completely unprecedented for such a high-level language (besides Rust), I just don't see a reason for this.

Kotlin relies on sandboxing for security. Yes, for languages like Rust and C that are used to write the sandbox and the OS themselves you have a point - but not for a language like Kotlin that is *made* to be sandboxed - Android is completely sandboxed, so are Wasm and Js; The JVM as a whole is a walking security nightmare; and Kotlin/Native exists in name only.

Immutable lists serve a different point in Kotlin: thread-safety and iteration-safety.

**Comment by Laxystem** [comment]
For example, when writing a game engine, I'm keeping vectors immutable and open; Yes, one can abuse that to render completely different things on my game's window - but that same malware can just create its own window, or encrypt your entire filesystem, because the JVM is unsecure af.

And this is without mentioning mixins that can just inject into your immutable list implementation at runtime anyway.

**Comment by yogurtearl** [comment]
Could have a sub-interface that is sealed. 

Then most people can use `ImmutableList` if they want to allow for alternate impls. 

But if you really want a restricted set of implementations, you can use `SealedImmutableList`

```kotlin
interface ImmutableList

sealed interface SealedImmutableList: ImmutableList

class PersistentList: SealedImmutableList { ... }
class SmallList: SealedImmutableList { ... }
```

**Comment by popematt** [comment]
If I am going to continue using `kotlinx.collections.immutable`, I need to be able to take a mutable list that I own and say that it is now immutable instead of having to make a defensive copies in the hot path of my code.

For example (pseudocode):
```kotlin
fun buildTree(treeReader: TreeReader): Tree? {
    return when (treeReader.nextNodeType()) {
        LEAF -> Tree.Leaf(treeReader.getCurrentValue())
        BRANCH -> {
            treeReader.stepDown()
            val children = mutableListOf<Tree>()
            while (true) {
                val child = buildTree(treeReader)
                child ?: break
                children.add(child)
            }
            treeReader.stepUp()
            // Defensive copy here would be prohibitively expensive.
            Tree.Branch(ImmutableListAdapter(children))
        }
        null -> null
    }
}
```

I just got a significant (~75%) reduction of the memory allocation rate in the hot path of a project I maintain by getting rid of redundant defensive copying.

----

> Could have a sub-interface that is sealed.

I think that @yogurtearl's suggestion is probably the best choice for balancing use cases like mine with the more security-focused use cases.

**Comment by Laxystem** [comment]
I agree with that. As for implementation, I would:

```kotlin
sealed interface ConstantCollection<T> : PersistentCollection<T>

fun <T> persistentListOf(vararg contents): ConstantList<T> // change return type

fun <T> List<T>.toConstantList(): ConstantList<T> = if (this is ConstantList) this else persistentListOf() + this // add new converters
```

Because `Constnat` implies completely-unchangeable, similarly to how `val` is immutable but can be changed via contract breach (mixins or wrong implementation).

**Comment by JakeWharton** [comment]
`val` means read only. Neither the reference it points to nor the returned reference itself has to be immutable.

```kotlin
class NotImmutable {
  val a: MutableList<Int> get() = ArrayList()
}
```

**Comment by pacher** [comment]
The way I see it, the source of the dispute is that we essentially have two libraries in one: immutable collections and persistent collections.
So we also have two types of users:
1. Users like @Sporking, who came for Guava replacement and are only interested in the immutable part. The library's implementation of immutable list is expected to be as optimal as it gets, and it is hard to imagine a use case where someone would want to have their own implementation (maybe @mcpiroman can share a use case ?).
2. Users who came for persistent collections and might want to have custom mutating implementations for special cases like @rhdunn

So basically, if we seal ImmutableList but keep PersistentList open, that should make everyone happy ;-) 
But if we keep ImmutableList open, it does probably make sense (as @yogurtearl suggested) to expose some subtype like `StrictImmutableList` with stronger guarantees for those who need it.

**Comment by popematt** [comment]
Regarding [my use case for open immutable collections](https://github.com/Kotlin/kotlinx.collections.immutable/issues/147#issuecomment-2266219287), it could also be solved if the library provided special builders for the immutable collections that avoided making defensive copies. The builder could guarantee that a mutable collection is wrapped in an immutable collection and that the mutable collection will not be leaked anywhere else. Something like this strawman example might work:
```kotlin
class ImmutableListBuilder<T> {
    private var size: Int = 0
    private val elements: Array<T?>? = null;

    /** Allocate new array if elements is null or if more capacity is needed */
    private fun ensureCapacity(n: Int): Array<T?> = TODO()

    fun add(element: T): ImmutableListBuilder<T> {
        val array = ensureCapacity(size + 1)
        array[size++] = element
        return this
    }

    /** Builds an ImmutableList and clears this builder */
    fun buildAndReset(): ImmutableList<T> {
        // Wrap the array without making a defensive copy
        val immutableList = ArrayBackedImmutableList(elements)
        // Ensure that only the ArrayBackedImmutableList has a reference to the backing array
        elements = null
        return immutableList
    }
}
```

**Comment by Laxystem** [comment]
> `val` means read only. Neither the reference it points to nor the returned reference itself has to be immutable.

I was referring to fieldful properties.

**Comment by DmitryNekrasov** [comment]
We discussed sealing while designing the ordered and unordered types in #364 and decided not to seal, since the 0.5.0 migration guide promises that the persistent interfaces may be implemented outside the library, and the new flavor interfaces are meant to be implementable too.
